### ISSUE TITLE
Download Strategy Resource Dependency Missing Cases

### ISSUE THREAD

**Issue by TheMDev** [Homebrew/brew#15395]
### `brew doctor` output

```shell
Your system is ready to brew.


brew doctor with the test script:


Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
  ?? Formula/test_rar.rb
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.0.17-44-gf7168bf-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: f7168bf234ee48a299b5eeb98e9895349bdcbf4a
Last commit: 3 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d2b110ec2585c7d2c81f1db9def3f87bfceed552
Core tap last commit: 2 hours ago
Core tap branch: master
Core tap JSON: 10 May 01:41 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.cNgVMNyavX/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 14.0.3 build 1403
Git: 2.40.1 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-x86_64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
```

### What were you trying to do (and why)?

I was originally unsure if this was a bug or not and had opened the discussion [here](https://github.com/orgs/Homebrew/discussions/4496#discussion-5178643) however, I am now fairly sure I have tracked down the issue. I originally tried to download a rar as a resource, but it errors out over missing the formulae for unrar. However, I do not want to extract the rar file, so there is no reason it should require unrar. I tried specifying the download strategy `using: :nounzip` however, this does not fix the issue. I have tried several different rar files unsuccessfully. I wrote a short test formula below using a rar file from archive.org.

### What happened (include all command output)?

`
~ % cat /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/test_rar.rb
`
```
class TestRar < Formula
  desc "test"
  homepage "test"
  url "https://archive.org/download/SpaceCadet_Plus95/Space_Cadet.rar", using: :nounzip
  version "test"
  sha256 "3cc5dfd914c2ac41b03f006c7ccbb59d6f9e4c32ecfd1906e718c8e47f130f4a"
  license "test"
end
```
`
~ % HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source --verbose --debug test_rar
`
```
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/test_rar.rb
Warning: No available formula with the name "unrar" (dependency of test_rar). Did you mean unar?
==> Searching for similarly named formulae and casks...
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unar.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromDefaultTapPathLoader): loading lunar
==> Formulae
unar

To install unar, run:
  brew install unar

==> Casks
lunar

To install lunar, run:
  brew install --cask lunar
```

### What did you expect to happen?

Specifying `using: :nounzip` should not require the dependencies to extract the downloaded file. Looking at [dependency_collector.rb:175](https://github.com/Homebrew/brew/blob/f7168bf234ee48a299b5eeb98e9895349bdcbf4a/Library/Homebrew/dependency_collector.rb#L175) it seems both `CurlPostDownloadStrategy` & `NoUnzipCurlDownloadStrategy` are missing from the `resource_dep` function. Changing the function to include them does seem to resolve the issue. However as I am still fairly new to the brew code base, I wanted to create an issue first to make sure my understanding of the function and my fix was correct before opening a pull request.

Any feedback is greatly appreciated.

```
  def resource_dep(spec, tags)
    tags << :build << :test
    strategy = spec.download_strategy

    if strategy <= HomebrewCurlDownloadStrategy
      @deps << curl_dep_if_needed(tags)
      parse_url_spec(spec.url, tags)
    elsif strategy <= NoUnzipCurlDownloadStrategy
      # allow no unzip to pass through
    elsif strategy <= CurlPostDownloadStrategy
      # allow POST to pass through
    elsif strategy <= CurlDownloadStrategy
      parse_url_spec(spec.url, tags)
    elsif strategy <= GitDownloadStrategy
      git_dep_if_needed(tags)
    elsif strategy <= SubversionDownloadStrategy
      subversion_dep_if_needed(tags)
    elsif strategy <= MercurialDownloadStrategy
      Dependency.new("mercurial", tags)
    elsif strategy <= FossilDownloadStrategy
      Dependency.new("fossil", tags)
    elsif strategy <= BazaarDownloadStrategy
      Dependency.new("breezy", tags)
    elsif strategy <= CVSDownloadStrategy
      cvs_dep_if_needed(tags)
    elsif strategy < AbstractDownloadStrategy
      # allow unknown strategies to pass through
    else
      raise TypeError, "#{strategy.inspect} is not an AbstractDownloadStrategy subclass"
    end
  end
```

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
Already shown in the What Happened section.
```

**Comment by Bo98** [comment]
The `CurlPostDownloadStrategy` behaviour seems correct (it'll use the `strategy <= CurlDownloadStrategy` case) as it's really the same as regular `CurlDownloadStrategy` for the tools needed to download and extract.

But I agree that `NoUnzipCurlDownloadStrategy` should be special cased.

You've also pointed out another issue here that even if `nounzip` wasn't used, the `unrar` dependency is incorrect as this doesn't exist anymore - so we should also correct that to depend on something else to extract the rar instead for formulae that need it.

**Comment by TheMDev** [comment]
Yeah looking through all the previous threads regarding rar I understand the licensing issues with it. However, I don't think there is really a good alternative for it after some testing I have found that while `p7zip` and `sevenzip` do extract a rar they throw `ERROR: Unsupported Method` for each file. `unar` does seem to work but is missing some functionality that doesn't make it a one to one replacement and there may also be some licensing issues. The `rar` cask seems like the obvious choice but I don't think we can include a cask as its not really supported. Would it just make more sense to have it throw an error or warning that says brew doesn't support extracting rar? Any feedback would be greatly appreciated.

**Comment by Bo98** [comment]
Yes, it's a legal mess.

`unar` or `libarchive` (`bsdtar`, depsite the name, supports rar) are options. `libarchive` is quite nice as macOS ships it, but it's currently too old (needs 3.6.0 or later), but that may change in a future macOS as Apple do keep it relatively up-to-date.

I'm not sure how "good" of a support each offer in terms of historical versions, but rars seem rare enough to probably not matter.

**Comment by MikeMcQuaid** [comment]
> However as I am still fairly new to the brew code base, I wanted to create an issue first to make sure my understanding of the function and my fix was correct before opening a pull request.

This seems like a reasonable first attempt, can you open a PR and we can comment there?



> `unar` or `libarchive` (`bsdtar`, depsite the name, supports rar) are options. I'd say `libarchive` is the most desirable option as macOS ships it, but it's currently too old (needs 3.6.0 or later). That may change in a future macOS as Apple do keep it relatively up-to-date, which will eventually mean no dependency required for those versions.

Fine with any of these 👍🏻

**Comment by danielnachun** [comment]
I believe `archiver` also supports extracting `rar` files and has the potential advantage of having no runtime dependencies because it's written in `go`.  Actually if `archiver` supports extracting most of the archive formats we allow, I wonder if we could eventually simplify some of our unarchiving code by relying on `archiver` as a more universal unarchiver than the more complicated mix of tools we use now.   This is certainly not a critical improvement at the moment but might be worth keeping in mind for future development.

**Comment by MikeMcQuaid** [comment]
@danielnachun I think it'd be good to rely on a longer-running existing formula for this. Definitely don't think we want a new universal archiver.

**Comment by TheMDev** [comment]
https://github.com/Homebrew/brew/pull/15409
